src/queue.c

```diff
@@ -967,6 +975,63 @@ _dispatch_get_mach_host_port(void)
 }
 #endif
 
+#if DISPATCH_USE_THREAD_LOCAL_STORAGE
+#include <unistd.h>
+#include <sys/syscall.h>
+
+#ifdef SYS_gettid
+DISPATCH_ALWAYS_INLINE
+static inline pid_t
+gettid(void)
+{
+	return (pid_t) syscall(SYS_gettid);
+}
+#else
+#error "SYS_gettid unavailable on this system"
+#endif
+
+#define _tsd_call_cleanup(k, f)  do { \
+		if (f && tsd->k) ((void(*)(void*))(f))(tsd->k); \
+    } while (0)
+
+static void
+_libdispatch_tsd_cleanup(void *ctx)
+{
+	struct dispatch_tsd *tsd = (struct dispatch_tsd*) ctx;
+
+	_tsd_call_cleanup(dispatch_queue_key, _dispatch_queue_cleanup);
+	_tsd_call_cleanup(dispatch_frame_key, _dispatch_frame_cleanup);
+	_tsd_call_cleanup(dispatch_cache_key, _dispatch_cache_cleanup);
+	_tsd_call_cleanup(dispatch_context_key, _dispatch_context_cleanup);
+	_tsd_call_cleanup(dispatch_pthread_root_queue_observer_hooks_key,
+			NULL);
+	_tsd_call_cleanup(dispatch_defaultpriority_key, NULL);
+#if DISPATCH_PERF_MON && !DISPATCH_INTROSPECTION
+	_tsd_call_cleanup(dispatch_bcounter_key, NULL);
+#endif
+#if DISPATCH_LOCK_USE_SEMAPHORE_FALLBACK
+	_tsd_call_cleanup(dispatch_sema4_key,
```

Comment:
I don't see what you are requesting here. You want the _tsd_call_cleanup to do the cast
to a ((struct dispatch_tsd*) ctx)  to obtain the equivalence of the tsd?